tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -5668,6 +5669,10 @@ void CGMSHLSLRuntime::EmitHLSLOutParamConversionInit(
     if (Param->isModifierOut()) {
       castArgList.emplace_back(tmpLV);
       castArgList.emplace_back(argLV);
+      if (isVector && !hlsl::IsHLSLVecType(argType)) {
+        DiagnosticsEngine &Diags = CGM.getDiags();
+        Diags.Report(Param->getLocation(), diag::warn_hlsl_implicit_vector_truncation);
+      }
```

Comment:
This is the only place where the inout modifier information and location information was all available. For output params, these casts can only be implicit. Trying to explicitly cast arguments fails to match the functions.

Reply:
> Trying to explicitly cast arguments fails to match the functions.
I'm not sure I understand why this is.  Is it because the explicit cast produces an RValue?  I don't think this should be by design, as LValue cast is allowed in fxc, I believe, but we do have problems supporting it.  So perhaps this is a side effect of our other issues preventing us from supporting LValue cast.

Reply:
The result is a failure to find the corresponding function:
```
 error: no matching function for call to 'TakeItOut'
```
 It is, as you say, because the parameter is an out and the cast produces something not an LValue. See https://github.com/microsoft/DirectXShaderCompiler/blob/a8d9780046cb64a1cea842fa6fc28a250e3e2c09/tools/clang/lib/Sema/SemaOverload.cpp#L4961

I think I'll add a comment where I depend on this behavior so we can catch it should we change it.